lib/pdn/pdn.c

```diff
@@ -0,0 +1,768 @@
+
+	/* Correct responses are at least 14 characters long
+	 * skip all shorter strings
+	 */
+	if ((index < 0) || (response == NULL) || (strlen(response) < 14)) {
+		return;
+	}
+
+	struct pdn_connection *const pdn_context =
+		(struct pdn_connection *)&pdn_contexts[index];
+
+	if (strstr(response, "CNEC_ESM") != NULL) {
+		/* AT event: +CNEC_ESM: <cause>,<cid> */
+		u8_t nw_error = 0;
+
+		sub_str = strchr(response, ':');
+		if (sub_str != NULL) {
+			nw_error = atoi(sub_str + 2);
+		}
+
+		sub_str = strchr(sub_str, ',');
+		if (sub_str != NULL) {
+			cid = atoi(sub_str + 1);
+		}
+
+		/* Check if context id match */
+		if (cid == pdn_context->context_id) {
+			/* Store network failure cause */
+			pdn_context->nw_error = nw_error;
+			event = PDN_EVENT_NW_ERROR;
+		}
+	} else if (strstr(response, "CGEV: ME PDN") != NULL) {
+		/* AT event: +CGEV: ME PDN ACT <cid> */
+		sub_str = strstr(response, "ACT ");
+		if (sub_str != NULL) {
+			cid = atoi(sub_str + 4);
+			/* Update if context id is correct */
+			if (cid == pdn_context->context_id) {
+				/* PDN deactivated */
+				pdn_context->state = PDN_STATE_IDLE;
+				event = PDN_EVENT_DEACTIVATED;
+			}
+		}
+	}
+
+	if ((event != PDN_EVENT_NA) && (pdn_context->callback != NULL)) {
+		pdn_context->callback(index, event);
+	}
+}
+
+static int local_context_reserve(const enum pdn_inet_type family)
+{
+	/* Search free PDN context index */
+	int i;
+
+	for (i = 0; i < CONFIG_PDN_NUM_CONTEXTS; i++) {
+		struct pdn_connection *const pdn_context = &pdn_contexts[i];
+
+		if (pdn_context->state == PDN_STATE_FREE) {
+			/* Clear any existing data */
+			memset(pdn_context, 0, sizeof(struct pdn_connection));
+
+			pdn_context->state = PDN_STATE_ALLOCATED;
+			pdn_context->family = (family == PDN_INET_DEFAULT) ?
+						      PDN_INET_IPV4V6 :
+						      family;
+			/* Set context id to invalid */
+			pdn_context->context_id = PDN_CID_INVALID;
+			/* Set PDN to indicate invalid/unread value. */
+			pdn_context->pdn_id = PDN_ID_INVALID;
+
+			/* Register AT event handler */
+			at_notif_register_handler((void *)PDN_IDX_TO_HANDLE(i),
+						  at_notif_handler);
+
+			return i;
+		}
+	}
+
+	return -ENOMEM;
+}
+
+static void local_context_release(const int index)
+{
+	struct pdn_connection *const pdn_context = &pdn_contexts[index];
+
+	/* Remove AT event handler */
+	at_notif_deregister_handler((void *)PDN_IDX_TO_HANDLE(index),
+				    at_notif_handler);
+
+	/* Mark to free */
+	pdn_context->callback = NULL;
+	pdn_context->state = PDN_STATE_FREE;
+	pdn_context->context_id = PDN_CID_INVALID;
+	pdn_context->pdn_id = PDN_ID_INVALID;
+}
+
+static int command_response_parse(struct pdn_connection *const pdn_context,
+				  char *const resp,
+				  const enum pdn_at_evt evt_id)
+{
+	const char *compare_str = pdn_at_evts_list[evt_id].at_str;
+	const int compare_length = strlen(compare_str);
+
+	if (strlen(resp) >= compare_length) {
+		if (strncmp(compare_str, resp, compare_length) == 0) {
```

Comment:
Why not simply call `strcmp(compare_str, resp)`? I don't see a reason for explicit length comparison here.

Reply:
This checks that received AT response match to sent command.  This method is called only when response value is expected (query), OK/ERROR is handled already by the at_cmd library.
Hence, AT response must be always longer than compare_str (expected start of the response) and strcmp will return negative values if start of the strings is same (length is not same => 0 is not returned). strcmp might also returns negative value event the strings are not same:
```
strcmp("+CCLK:", "+CFUN: 4") = -1
```
So I cannot be sure that reponse is correct if len is not taken into account.

And generally strncmp is safer e.g. null termination is missing for some reason.

Reply:
Ok I see the point with a longer response, could possibly use some comment though.

Missing NULL termination is not a valid argument here, both strings were just checked with `strlen`.

Reply:
I will add comment.